docs/topics/email.txt

```diff
@@ -27,6 +27,28 @@ In two lines::
         fail_silently=False,
     )
 
+Sending HTML emails (with required plain text part)::
+
+    from django.core.mail import EmailMultiAlternatives
+    from django.template.loader import render_to_string
+
+    # Create email object
+    msg = EmailMultiAlternatives(
+        "Subject here",
+        "Plain text content",
+        "from@example.com",
+        ["to@example.com"],
+    )
+
+    # Render HTML content from template and attach it to mail object
+    html_content = render_to_string(
+        "templates/emails/my_email.html",
+        context={"my_variable": 42},
+    )
+    msg.attach_alternative(html_content, "text/html")
+
+    msg.send()
```

Comment:
**If** my reading of things is right, this is actually the more slightly more complicated way of doing it? The `send_mail` function gained an `html_message` parameter in [`1.7`](https://docs.djangoproject.com/en/5.1/releases/1.7/#email) by the look of it (see [ticket 20817](https://code.djangoproject.com/ticket/20817)), so I _think_ this could be something like:

```python
from django.core.mail import send_mail

rendered_template = render_to_string(
    "templates/emails/my_email.html",
    context={"my_variable": 42},
)

send_mail(
    "Subject here",
    "Plain text content",
    "from@example.com",
    ["to@example.com"],
    html_message=rendered_template,
)

Reply:
@GitRon @kezabelle Hi! Thank you for this work and initial review.

I wonder why, then, when `PasswordResetForm` isolated email sending in `send_email`, this approach of using `html_message` was changed to use the "longer" approach: https://github.com/django/django/commit/a00b78b1e2a

Any ideas?

Reply:
@nessita and @kezabelle `send_mail()` and other functional APIs are "frozen" and new features show up only in `EmailMessage` and `EmailMultiAlternatives`. I don't know the exact history, but the intent seems to be that new code should use the classes, though the old functional APIs are also still supported.

I think it's a great idea to include examples of both APIs at the top of the page.

Right now, the reader has to dig [quite a bit later](https://docs.djangoproject.com/en/5.1/topics/email/#the-emailmessage-class:~:text=Not%20all%20features%20of%20the%20EmailMessage%20class%20are%20available%20through%20the%20send_mail()%20and%20related%20wrapper%20functions.%20If%20you%20wish%20to%20use%20advanced%20features%2C%20such%20as%20BCC%E2%80%99ed%20recipients%2C%20file%20attachments%2C%20or%20multi%2Dpart%20email%2C%20you%E2%80%99ll%20need%20to%20create%20EmailMessage%20instances%20directly.) in this page to find out that:

> Not all features of the EmailMessage class are available through the send_mail() and related wrapper functions. If you wish to use advanced features, such as BCC’ed recipients, file attachments, or multi-part email, you’ll need to create EmailMessage instances directly.

And the note box just below that mentions the "original functions" like `send_mail()` are retained "only for backwards compatibility. 

Those parts of the docs date back over a decade. They carefully avoid using the word "deprecated." But in [the django.core.mail code](https://github.com/django/django/blob/cdcd604ef8f650533eff6bd63a517ebb4ffddf96/django/core/mail/__init__.py#L78-L79) you'll find this limitation on updating `send_mail()` (and `send_mass_mail()`):

```py
def send_mail(...):
    """
    ...

    Note: The API for this method is frozen. New code wanting to extend the
    functionality should use the EmailMessage class directly.
    """
```

Reply:
@GitRon 

General 👍. A few thoughts:

- Maybe add something not available in `send_mail()` to the class example, like custom headers?
  `headers={"List-Unsubscribe": "<mailto:unsub@example.com>"},`. (To @kezabelle's point about send_mail() allowing html_message. A cc or bcc or attachment could also work.)
- Maybe change the heading to "Quick examples" (plural) and update the text to clarify the purpose of each example and link to the detailed docs. Something like:
  - "Use [`send_mail()`](#) for simpler cases. For example, to send a plain text only message:"  (Also, "In two lines" hasn't been accurate since blacken-docs.)
  - "Use [`EmailMessage`](#) or [`EmailMultiAlternatives`](#) for additional features. For example, to send an html + plain text message using a template and custom headers:"
- For the template rendering example, I'd lean toward rendering both plaintext and html from templates using the same context. (Or not including the plaintext body.) I think it's helpful for even minimal examples to encourage best practices. A lot of ESPs treat big differences in the text content of the two bodies as a spam signal.

Reply:
Hi @medmunds!

Thx so much for your feedback.

> Maybe add something not available in send_mail() to the class example, like custom headers?

👍 

> Maybe change the heading to "Quick examples" (plural) and update the text to clarify the purpose of each example and link to the detailed docs.

👌 

> For the template rendering example, I'd lean toward rendering both plaintext and html from templates using the same context. 

You are totally right. But I'd have still two templates, a txt- and an HTML-based one, right? That was one of the drivers for my pony express package 😅 

---

I've pushed a new commit.